### PR TITLE
Make Blockstore::get_slots_since() filter out results for dead slots

### LINKED ISSUE #28343: Panic from attempting to replay a dead slot
#### Problem

On a node running v1.14 (issue present in master too), we observed a panic shortly after the node fired up ReplayStage. The following log lines with annotations tell the story of what is happening and why we end with a panic:
```
// Validator process starts
[2022-10-07T20:05:04.522762389Z INFO  solana_validator] Starting validator with: ArgsOs {

// Ledger processing begins for what is already present
[2022-10-07T20:10:04.332963953Z INFO  solana_ledger::blockstore_processor] Processing ledger from slot 154185240...
[2022-10-07T20:10:04.333307300Z INFO  solana_ledger::blockstore_processor] ledger holds data through slot 154185596

// 154185568 bank created
[2022-10-07T20:12:35.339310609Z INFO  solana_metrics::metrics] datapoint: bank-new_from_parent-heights slot=154185568i block_height=139237307i parent_slot=154185563i ...

// 154185568 fails to verify and is marked dead
[2022-10-07T20:12:36.251737344Z WARN  solana_ledger::blockstore_processor] slot 154185568 failed to verify: failed to load entries, error: blockstore error

// Ledger processing finished, 154185563 still an active bank
[2022-10-07T20:12:42.324282843Z INFO  solana_ledger::blockstore_processor] ledger processed in 2 minutes, 37 seconds, 990 ms, 457 µs and 693 ns. root slot is 154185555, 32 banks: 154185555, 154185556, 154185557, 154185558, 154185559, 154185560, 154185561, 154185562, 154185563, 154185564, 154185565, 154185566, 154185567, 154185569, 154185570, 154185571, 154185572, 154185573, 154185574, 154185575, 154185576, 154185577, 154185578, 154185579, 154185580, 154185581, 154185582, 154185583, 154185584, 154185585, 154185586, 154185587

// set_hash error (this is a subroutine of bank creation for 154185568)
[2022-10-07T20:12:42.373863730Z ERROR solana_runtime::accounts_db] set_hash: already exists; multiple forks with shared slot 154185568 as child (parent: 154185563)!?

// 154185568 bank created second time since its' parent (154185563) not yet rooted so found by generate_new_bank_forks()
[2022-10-07T20:12:42.374852562Z INFO  solana_metrics::metrics] datapoint: bank-new_from_parent-heights slot=154185568i block_height=139237307i parent_slot=154185563i ...

// ReplayStage rejects 154185568 since we had previously marked it dead
[2022-10-07T20:12:42.610532191Z ERROR solana_metrics::metrics] datapoint: replay-stage-mark_dead_slot error="error: FailedToLoadEntries(DeadSlot)" slot=154185568i

// Debug panic that was put in to provide more information than the eventual panic that we would hit
[2022-10-07T20:12:56.778393865Z INFO  solana_runtime::accounts_db] remove_dead_slots_metadata: 1 dead slots, dead_slots: [154185568]
thread 'solBgAccounts' panicked at 'cleaning slot that was duplicate: 154185568', runtime/src/accounts_db.rs:7590:21
stack backtrace:
[2022-10-07T20:12:56.798757132Z INFO  solana_metrics::metrics] datapoint: Gossip streamer-send-sample_duration_ms=393i streamer-send-host_count=1164i streamer-send-bytes_total=2180067i streamer-send-pkt_count_total=2060i streamer-send-host_bytes_min=1050i streamer-send-host_bytes_max=9536i streamer-send-host_bytes_mean=1875i streamer-send-host_bytes_90pct=3179i streamer-send-host_bytes_50pct=1060i streamer-send-host_bytes_10pct=1060i
[2022-10-07T20:12:56.798764606Z INFO  solana_streamer::streamer] streamer send Gossip hosts: count:1164 [(69.197.13.5, SendStats { bytes: 9531, count: 9 }), (65.109.52.186, SendStats { bytes: 9531, count: 9 }), (65.108.127.108, SendStats { bytes: 8472, count: 8 }), (66.23.234.14, SendStats { bytes: 7413, count: 7 }), (65.108.226.27, SendStats { bytes: 7413, count: 7 })]
   0: rust_begin_unwind
             at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
   2: solana_runtime::accounts_db::AccountsDb::purge_slot_cache_pubkeys
   3: solana_runtime::accounts_db::AccountsDb::purge_slot_cache
   4: solana_runtime::accounts_db::AccountsDb::purge_slot
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2022-10-07T20:12:56.820420647Z ERROR solana_metrics::metrics] datapoint: panic program="validator" thread="solBgAccounts" one=1i message="panicked at 'cleaning slot that was duplicate: 154185568', runtime/src/accounts_db.rs:7590:21" location="runtime/src/accounts_db.rs:7590:21" version="\"1.14.6 (src:devbuild; feat:940802714)\""
```
So, basically, the issue is that the slot is getting marked dead in startup processing, but then also being processed by ReplayStage a second time. ReplayStage can handle playing dead slots; however, since this is the second time the slot was created, we encountered that `set_hash`, things were already in a bad state.

#### Proposed Solution
A couple options I see:
- Make the blockstore function that `ReplayStage::generate_new_bank_forks()` calls filter out dead slots
- Do more cleanup on the dead slot in initial processing such that creating & dropping the bank the second time doesn't error

#### Full log
[bug.log](https://github.com/solana-labs/solana/files/9751424/bug.log)

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
This function is used by ReplayStage to find new slots to replay. If we know the slot is dead, then we can ignore the slot instead of creating the bank and later finding out that it is dead.

Moreso, if the slot was processed in validator startup processing (`blockstore_processor::load_frozen_forks()`) and then also processed by `ReplayStage`, we will have created a bank for that slot twice, which causes a downchain failure.

#### Summary of Changes
Filter out dead slots from `Blockstore::get_slots_since()`.

Fixes #28343

### CODE REVIEW

**Comment by steviez** [comment]
Think this one needs to go in v1.14 and v1.13/v1.10

File: ledger/src/blockstore.rs [hunk 1]
```diff
@@ -2975,13 +2975,31 @@ impl Blockstore {
...
     pub fn get_slots_since(&self, slots: &[u64]) -> Result<HashMap<u64, Vec<u64>>> {
-        let slot_metas: Result<Vec<Option<SlotMeta>>> =
-            self.meta_cf.multi_get(slots.to_vec()).into_iter().collect();
+        let dead_slots: Result<Vec<Option<bool>>> = self
+            .dead_slots_cf
+            .multi_get(slots.to_vec())
+            .into_iter()
+            .collect();
+        let dead_slots = dead_slots?;
+        // If there is anything in the dead_slots cf for this slot, it is dead
```
Comment (jeffwashington) [inline ledger/src/blockstore.rs:2987]: cf = column family?

File: ledger/src/blockstore.rs [hunk 2]
```diff
@@ -2975,13 +2975,31 @@ impl Blockstore {
...
+        let alive_slots: Vec<Slot> = slots
+            .iter()
+            .zip(dead_slots)
+            .filter(|(_slot, dead)| dead.is_none())
+            .map(|(slot, _dead)| *slot)
+            .collect();
+
+        let slot_metas: Result<Vec<Option<SlotMeta>>> = self
+            .meta_cf
+            .multi_get(alive_slots.to_vec())
```
Comment (jeffwashington) [inline ledger/src/blockstore.rs:2997]: alive_slots is already a vec. is this required to clone it?

File: ledger/src/blockstore.rs [hunk 3]
```diff
@@ -5705,6 +5723,7 @@ pub mod tests {
 
     #[test]
     fn test_get_slots_since() {
+        solana_logger::setup();
```
Comment (jeffwashington) [inline ledger/src/blockstore.rs:5726]: you can get rid of this

File: ledger/src/blockstore.rs [hunk 4]
```diff
@@ -2975,13 +2975,31 @@ impl Blockstore {
...
+
+        let slot_metas: Result<Vec<Option<SlotMeta>>> = self
+            .meta_cf
+            .multi_get(alive_slots.to_vec())
+            .into_iter()
+            .collect();
         let slot_metas = slot_metas?;
 
-        let result: HashMap<u64, Vec<u64>> = slots
+        let result: HashMap<u64, Vec<u64>> = alive_slots
```
Comment (jeffwashington) [inline ledger/src/blockstore.rs:3002]: `u64` could be `Slot` for clarity and consistency.

**Review by jeffwashington — APPROVED** [review]
lgtm. @carllin should review the meaning/concept of `get_slots_since` not returning dead slots for correctness.

**Comment by steviez** [comment]
Talking with Carl, this solution may not be the safest. I have another idea, but it is a little more involved. Will pursue that

**Comment by steviez** [comment]
Circling back, this PR is NOT the correct approach to take as this comment calls out
https://github.com/solana-labs/solana/blob/6972afe2c4ed4202268dd71dc7d223f8348b4d01/core/src/replay_stage.rs#L1747-L1749

and which is later shown here

https://github.com/solana-labs/solana/blob/6972afe2c4ed4202268dd71dc7d223f8348b4d01/core/src/cluster_slot_state_verifier.rs#L499-L515

ReplayStage needs to know about these dead slots, so changing the function to not return them will definitely break things. Abandoning this PR in lieu of another apporach

**Pull request closed without merging**